src/frontend/src/helpers/websocket.ts

```diff
@@ -0,0 +1,411 @@
+
+  constructor(connection: WSConnection, storage: Storage) {
+    this.connection = connection;
+    // this.synced = false;
+    this.connected = false;
+    this.failed = false;
+    // this.isSyncing = false;
+    this.offlineMode = 0; // TODO change this to look up the cookie
+    this.timeoutCount = 0;
+    this.timeoutInterval = null;
+    this.key = 0;
+    this.storage = storage;
+    this.socket = new SeashellWebsocket(this.connection.wsURI, this.connection.publicKey);
+    this.callbacks = [];
+
+    // These functions are not available in offline mode.
+    this.ping = this.make_offline_disabled('ping');
+    this.newProject = this.make_offline_disabled('newProject');
+    this.newProjectFrom = this.make_offline_disabled('newProjectFrom');
+    this.deleteProject = this.make_offline_disabled('deleteProject');
+    this.restoreFileFrom = this.make_offline_disabled('restoreFileFrom');
+    this.getUploadFileToken = this.make_offline_disabled('getUploadFileToken');
+    this.getExportToken = this.make_offline_disabled('getExportToken');
+    this.marmosetSubmit = this.make_offline_disabled('marmosetSubmit');
+    // These functions do nothing and just resolve in offline mode.
+    this.lockProject = this.make_offline_noop('lockProject');
+    this.forceLockProject = this.make_offline_noop('forceLockProject');
+    this.unlockProject = this.make_offline_noop('unlockProject');
+    this.archiveProjects = this.make_offline_noop('archiveProjects');
+    
+    // These functions either:
+    //  - return the online result if online.
+    //  - return the offline result if offline.
+    this.getProjects = this.make_offline_enabled('getProjects');
+    this.listProject = this.make_offline_enabled('listProject');
+    this.readFile = this.make_offline_enabled('readFile');
+    this.getFileToRun = this.make_offline_enabled('getFileToRun');
+    this.getSettings = this.make_offline_enabled('getSettings');
+    // this.getMostRecentlyUsed = this.make_offline_enabled('getMostRecentlyUsed');
+
+    // These functions:
+    //  - invoke the offline version if offline
+    //  - invoke the both the offline version and the online
+    //    version if online, returning the offline version.
+    //  - marks the local storage as dirty and queues up a sync
+    this.newFile = this.make_offline_enabled('newFile', true);
+    this.writeFile = this.make_offline_enabled('writeFile', true);
+    this.deleteFile = this.make_offline_enabled('deleteFile', true);
+    this.renameFile = this.make_offline_enabled('renameFile', true);
+    this.saveSettings = this.make_offline_enabled('saveSettings', true);
+    this.setFileToRun = this.make_offline_enabled('setFileToRun', true);
+    // this.updateMostRecentlyUsed = this.make_offline_enabled('updateMostRecentlyUsed', true)
+  }
+
+  // Connects the socket, sets up the disconnection monitor
+  public async connect(): Promise<void> {
```

Comment:
Probably want to design this function so that if user is disconnected, call this function directly to reconnect.